src/common/partitioning/src/lib.rs

```diff
@@ -0,0 +1,90 @@
+use std::{any::Any, sync::Arc};
+
+use common_error::DaftResult;
+use futures::stream::BoxStream;
+use serde::{Deserialize, Serialize};
+#[cfg(feature = "python")]
+use {
+    common_py_serde::{deserialize_py_object, serialize_py_object},
+    pyo3::PyObject,
+};
+
+/// Common trait interface for dataset partitioning, defined in this shared crate to avoid circular dependencies.
+/// Acts as a forward reference for concrete partition implementations. _(Specifically the `MicroPartition` type defined in `daft-micropartition`)_
```

Comment:
still a little bit confused why this is a trait instead of us just being able to use MicroPartition directly. Maybe explain this more? Or are we expecting to expand more. add to docs maybe

Reply:
It's pretty much the equivalent of a [forward declaration](https://www.learncpp.com/cpp-tutorial/forward-declarations/). In `daft-logical-plan`, we don't have a concept of `MicroPartition` yet, and we generally dont want to because 'daft-micropartition' is a pretty heavy crate that we don't want part of the logical plan. We did a similar pattern with `Expr::Subquery`